lib/time/time.go

```diff
@@ -0,0 +1,548 @@
+// It is concurrency-safe and idempotent.
+func LoadModule() (starlark.StringDict, error) {
+	return starlark.StringDict{
+		ModuleName: Module,
+	}, nil
+}
+
+// NowFunc is a function that generates the current time. Intentionally exported
+// so that it can be overridden, for example by applications that require their
+// Starlark scripts to be fully deterministic.
+var NowFunc = time.Now
+
+func parseDuration(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var d Duration
+	err := starlark.UnpackArgs("duration", args, kwargs, "d", &d)
+	return d, err
+}
+
+func parseLocation(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var s string
+	if err := starlark.UnpackArgs("location", args, kwargs, "s", &s); err != nil {
+		return nil, err
+	}
+	loc, err := time.LoadLocation(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return starlark.String(loc.String()), nil
+}
+
+// SleepFunc is a function that pauses the current goroutine for at least d.
+// Intentionally exported so that it can be overridden, for example by
+// applications that require their Starlark scripts to be fully deterministic.
+var SleepFunc = time.Sleep
+
+func sleep(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var dur Duration
+	if err := starlark.UnpackPositionalArgs("sleep", args, kwargs, 1, &dur); err != nil {
+		return starlark.None, err
+	}
+
+	SleepFunc(time.Duration(dur))
+	return starlark.None, nil
+}
+
+func parseTime(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var (
+		x, location string
+		format      = time.RFC3339
+	)
+	if err := starlark.UnpackArgs("time", args, kwargs, "x", &x, "format?", &format, "location", &location); err != nil {
+		return nil, err
+	}
+
+	if location == "" {
+		t, err := time.Parse(format, x)
+		if err != nil {
+			return nil, err
+		}
+		return Time(t), nil
+	}
+
+	loc, err := time.LoadLocation(location)
+	if err != nil {
+		return nil, err
+	}
+	t, err := time.ParseInLocation(format, x, loc)
+	if err != nil {
+		return nil, err
+	}
+	return Time(t), nil
+}
+
+func fromUnixTime(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	var x int
+	if err := starlark.UnpackPositionalArgs("from_timestamp", args, kwargs, 1, &x); err != nil {
+		return nil, err
+	}
+	return Time(time.Unix(int64(x), 0)), nil
+}
+
+func now(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	return Time(NowFunc()), nil
+}
+
+// Duration is a Starlark representation of a duration.
+type Duration time.Duration
+
+func newDuration(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
```

Comment:
This function looks unnecessary. It merely asserts that its argument is a Duration, and returns it unchanged.

Reply:
Hmmm if I understand the code correctly it calls the `Unpack()` function of `Duration` below to construct a new duration. All the type handling/conversion/construction etc. is done below, isn't it?
What is the correct way in your opinion to construct a new duration?

Reply:
@alandonovan I have the same question as @srebhan, to me this is a constructor that leverages the `Unpack` interface. Would love advice on an alternative